### ISSUE TITLE
Design confusion in demobench and cash app w.r.t. usage of network service tagging

### ISSUE THREAD

**Issue by gendal** [corda/corda#567]
I performed this experiment using a DemoBench from the M11 snapshot.   I observed this in M10 also so it's a pre-existing problem, not a regression.

Reproduction Steps:

1) Use DemoBench to create and start a node, configuring it with the cash service as well as the three currency issuing services.

2) Open explorer and try to issue cash to somebody. Observe that only one currency is offered (usually CHF in my case)

3) Look at node logs; observe that the node THINKS it is configured with the ability to issue in the three currencies.

```
[INFO ] 2017-04-20T18:30:42,041 [main] internal.Node.registerFlowInitiator - Reg
istering flow net.corda.flows.ContractUpgradeFlow$Instigator
[INFO ] 2017-04-20T18:30:42,299 [main] BasicInfo.printBasicNodeInfo - Providing 
network services              : corda.cash, corda.issuer.CHF, corda.issuer.GBP, 
corda.issuer.USD
[INFO ] 2017-04-20T18:30:42,300 [main] BasicInfo.printBasicNodeInfo - Loaded plu
gins                          : net.corda.bank.plugin.BankOfCordaPlugin

```

Observe that the node creates keys for each issued currency:

```
Richard-MacBook-Pro:bankofrichard richard$ ls
artemis				dependencies
attachments			identity-private-key
certificates			identity-public
corda.cash-private-key		logs
corda.cash-public		node.conf
corda.issuer.CHF-private-key	persistence.mv.db
corda.issuer.CHF-public		plugins
corda.issuer.GBP-private-key	process-id
corda.issuer.GBP-public		shell-commands
corda.issuer.USD-private-key	version
corda.issuer.USD-public
```

It COULD be a bug in the explorer, I guess... but nothing in the explorer logs

**Comment by mikehearn** [comment]
The problem is in the cash app, and the way the code is written more generally. We need to sit down and solve the underlying conceptual confusion, as it will show up as we change the identity layer.

The ability to issue cash in a currency is being triggered by telling the node to advertise a "network service". Originally a network service was simply a string, advertised in the network map. The presence of a string was intended to flag support for flows that might be useful to anyone on the network, like if you are an oracle.

When notaries were integrated the concept became more complicated, with the result that a service now has a distinct identity and set of identity keys. This is because a node is always owned by a specific legal entity, but a notary is a kind of distinct "virtual entity" that doesn't map to anyone in particular but rather, represents a kind of cooperative.

Issuance in the current cash app is (for some reason, I don't remember doing this) tied up with the notion of network services. This is not correct. Firstly, it makes no conceptual sense - issuing money isn't a service you just offer to anyone who turns up.  Secondly, no actual separate identities are provided, so they end up with the same legal name, which confuses things. This code in the explorer then results in the multiple issuers being eliminated:

```
        issuerChoiceBox.apply {
            items = issuers.map { it.legalIdentity }.unique().sorted()
            converter = stringConverter { it.name }
            visibleProperty().bind(transactionTypeCB.valueProperty().map { it == CashTransaction.Pay })
        }
```

Note the `.unique().sorted()` call, which is why CHF comes first.

It doesn't make sense for cash issuers to be a network service. Instead, enabling the cash app should simply enable you to issue any currency you like in whatever quantities you like. The cash app would later be upgraded to understand the concept of limiting exposure to issuers.

This leads to something that's been bugging me for a while which I have not yet attempted to address: the DemoBench lets you pick what services you wish the node to expose. That also makes no sense. The only thing you should be able to enable/disable is the CorDapps that are loaded. Services are provided by CorDapps, but currently you can start a node with no apps in the list, and some services selected, which is conceptually wrong except for a few 'special' things like being a notary.

So unfortunately this is not a quick fix. Instead we have to untangle some concepts that got tangled together due to being underspecified. I need to formally define what a network service is, in the docs, and then make sure everyone understands it. We need to split cash out into being a real cordapp instead of its current privileged position of being a "platform app" i.e. neither platform nor app. And the we need to adjust DemoBench to come with some useful apps like cash bundled, and let you enable or disable them before you start a node.

In short ... there's work to do.

**Comment by chrisr3** [comment]
> DemoBench lets you pick what services you wish the node to expose

IIRC, this seemed like the "obvious" way to configure the `extraAdvertisedServiceIds` element in `node.conf`, e.g.:
```json
"extraAdvertisedServiceIds" : [
    "corda.cash",
    "corda.issuer.GBP"
]
```
or
```json
"extraAdvertisedServiceIds" : [
    "corda.notary.validating"
],
```
> The only thing you should be able to enable/disable is the CorDapps that are loaded.

So does this mean that we need a separate BankOfCorda CorDapp for each currency? And that DemoBench needs to "service load" each CorDapp in order to extract its list of services for writing into `node.conf`? I previously experimented with service loading CorDapps so that DemoBench could determine the correct set of `User` permissions to use (prior to the creation of the "ALL" permission). This required that DemoBench always be able to execute the `CordaPluginRegistry` class's no-arg constructor.

**Comment by gendal** [comment]
Mike - thanks for the write-up; very helpful.

**Comment by mikehearn** [comment]
> So does this mean that we need a separate BankOfCorda CorDapp for each currency? 

No. The cash app should let you issue any currency. And the "Bank of Corda" app should probably go away or just become a really thin wrapper around the cash app.

> And that DemoBench needs to "service load" each CorDapp in order to extract its list of services for writing into node.conf? 

No. Apps should not be triggered by services appearing in the config file. Services should be triggered by apps, with the exception of one or two special services like notaries (which can have their own config sections enabling them anyway). So we should get away from specifying services in the config file.

**Comment by joeldudleyr3** [comment]
This now works correctly, possibly due to the removal of the concept of advertised services.
